### ISSUE TITLE
jumps 10 min. up to 1 hour; compared to 12 months ago

### ISSUE THREAD

**Issue by jzacsh** [restic/restic#1098]
**tl;dr** I'm just looking for feedback before I find myself doing `git bisect` this weekend to narrow this down...

## Output of `restic version`

```
$ /home/jzacsh/media/src/restic/bin/restic version
restic 0.7.0 (v0.7.0-21-g357e2e40)
compiled with go1.8.3 on linux/amd64
```

## How did you start restic exactly? (Include the complete command line)
```
$ cat ~/.config/restic-exclude
$HOME/back/local
$HOME/.android/
$HOME/.cinnamon/
$HOME/.cache/
$HOME/.config/google-chrome/
$HOME/.mozilla/firefox/
$HOME/.gnome/apps/
$HOME/.gradle/
$HOME/.npm/
$HOME/.thumbnails/
$HOME/.tor-browser*/
$HOME/.opam/
$HOME/.local/share/lxc/*/rootfs
$ANDROID_HOME/
$ANDROID_HOME/../studio
$HOME/media/src/linux-kernel/
$HOME/tmp/
$HOME/usr/local/vm/
$ /home/jzacsh/media/src/restic/bin/restic backup \
    --repo /home/jzacsh/back/local/restic \
    --exclude-file /home/jzacsh/.config/restic-exclude \
    /home/jzacsh
```

## What backend/server/service did you use?
n/a; local filesystem, same as source (one disk, one partition)

## Expected behavior
Expected this to take 10 minutes. This *exact* command I have logs going back over a year for. 

## Actual behavior

Running time, since update to 0.7.0, is now an hour. I actually upgraded said machine because a *different* laptop I installed restic on the other day, was showing 1hour+ backup times, and I thought, _"huh, my old machine doesn't take that long"_ ... Well, **now** that machine takes an hour also :P 

## Steps to reproduce the behavior

I plan to checkout old versions, like v0.3.0, and see if I can narrow down when restic started taking so long. But I thought I'd just open this bug to have a discussion first, before I start doing a million checkouts...

## Do you have any idea what may have caused this?

I unfortunately didn't save the hash my git repo for restic was at, before I did a `git pull`, so I cannot tell you exactly which ref my old restic was at. _(I have logs of `restic version` output, which my script always executed, but back then restic just said, "unknown  version compiled with go 1.6")_. However, the timestamp on some of my directories in my git repo say "Jun 2016", so I'll pick old versions, based on that.

**Comment by fd0** [comment]
Hey, thanks for raising this issue. I've got a few ideas you can try:
 * Unfortunately you did not include the actual backup output for restic, can you do that? Is there a line `using parent snapshot [...]`?
 * Write a debug log and have a look for a few files, there should be information in the debug log which tells you what restic does. Usually it tries to check files based on timestamps and use the old content from the previous backup if the files haven't changed. Do you see that?
 * Use `git reflog` to see which commit was checked out before the last pull, so you have a starting point for the bisect

I suspect that the parent snapshot detection does not work (somehow), so restic re-reads and re-hashes all files (which, at least for a local backend, takes a lot longer). Can you observe high CPU load and much IO from restic?

**Comment by jzacsh** [comment]
@fd0, thanks for the quick response; some quick answers but I'll be back with more details in a bit...

> *   Unfortunately you did not include the actual backup output for restic, can you do that? Is there a line `using parent snapshot [...]`?
Yup, here's seems parent snapshots are detected, per the last two runs' logs:
```
+ /home/me/media/src/restic/bin/restic backup --repo /home/me/back/local/restic --exclude-file /home/me/.config/restic-exclude /home/me
using parent snapshot 6bce45bb
scan [/home/me]
[0:08] 47179 directories, 209123 files, 38.748 GiB
scanned 47179 directories, 209123 files in 0:08
[1:01:30] 100.00%  0B/s  38.748 GiB / 38.748 GiB  256302 / 256302 items  0 errors  ETA 0:00
duration: 1:01:30, 10.75MiB/s
snapshot 376a8058 saved


+ /home/me/media/src/restic/bin/restic backup --repo /home/me/back/local/restic --exclude-file /home/me/.config/restic-exclude /home/me
using parent snapshot 376a8058
scan [/home/me]
[0:08] 47179 directories, 209123 files, 38.748 GiB
scanned 47179 directories, 209123 files in 0:08
[1:01:34] 100.00%  0B/s  38.748 GiB / 38.748 GiB  256302 / 256302 items  0 errors  ETA 0:00

duration: 1:01:34, 10.74MiB/s
snapshot 1c77a790 saved
```

> Write a debug log and have a look for a few files, there should be information in the debug log which tells you what restic does. Usually it tries to check files based on timestamps and use the old content from the previous backup if the files haven't changed. Do you see that?

Perfect, this is running now. Hopefully this will show me something glaring/obvious before I start comparing versions. I'll **report back**  what I find in these logs.

> Use `git reflog` to see which commit was checked out before the last pull, so you have a starting point for the bisect

Great - this is a good starting point _(the first is a commit from August, so this is good enough - but I'm not sure if it's technically the *exact* ref I had before doing a `pull`, since that might have not caused a fast-forward?)_

```
20170713  357e2e40 (HEAD -> master, origin/master, origin/HEAD) HEAD@{0}: pull origin master: Fast-forward
20160816  812cb0ba HEAD@{1}: pull origin master: Fast-forward
20160720  e8e45fe2 HEAD@{2}: pull origin master: Fast-forward
20160717  b8c7622a HEAD@{3}: clone: from git://github.com/restic/restic.git
```
(I prepended dates to the lines...)

> I suspect that the parent snapshot detection does not work (somehow), so restic re-reads and re-hashes all files (which, at least for a local backend, takes a lot longer).

So, I guess is ^this ruled out by the above logs showing `using parent snapshot`?

> Can you observe high CPU load and much IO from restic?

I'll keep an eye out & **report back**. FWIW: on my weaker laptop _(not the one that had restic running in 10-min. backups for a year)_, where I _first_ got suspicious last week, a fresh restic backup crashed gnome and ground things to a halt ... but I'l deal with that case later.

> Does maybe your hostname change in between backups (sometimes a problem with Docker/rkt/kubertenes...)?

Nope, nothing so interesting on my machines.

**Comment by fd0** [comment]
This is a bit of a mystery, so I'm eager to find out what's going on here. Did you observe CPU load and/or IO?

**Comment by jzacsh** [comment]
@fd0 since these backups take so long, I let it run and forgot about it. I just came back tonight and found my `debug.log` is 112G. It filled up my local disk until things came to a halt & restic exited with a panic _"no space left on device"_ I'll dig in later this week and make sense of it, but if you'd like I'd be happy to chat 1:1 at some point over IRC/email with exact outputs from the file for you to see.

otherwise, I'll report back here if I start making sense of the debug.log, and have ideas of my own.

**Comment by fd0** [comment]
Oh wow, 112G of debug log for about 40G of data? Something is really off here! What's your timezone? I'm usually available in the evening (Berlin time, CEST, GMT+2) on IRC.

**Comment by fd0** [comment]
So, did you find out what went wrong? We've recently merged several fixes for similar things, can you try again with the latest master branch?

I'm going to close this issue for now as it is quite old, please feel free to add further comments (and then we can easily reopen the issue). Thanks!
